### PR TITLE
fix(core): cap SARSA array-loop sequence length before scan hang

### PR DESCRIPTION
**Pull request by rama0x1**
## Summary

`alberta_framework/core/sarsa.py`'s `run_sarsa_from_arrays` and
`run_sarsa_from_arrays_final_state` hand their pre-collected step arrays
(`observations`, `rewards`, `terminated`, `next_observations`) straight to
`jax.lax.scan` with no bound on the leading (step) axis and no check that
the four arrays share a length. A hostile or mistaken caller supplying a
huge array forces JAX to trace/compile a scan of that length, hanging the
process well before any step executes.

This is the same hang class just fixed for other scan-driven array loops in
this repo (`core/partner_policy_fusion.py` #1991, `utils/nexting.py` #1992,
`core/learners.py` #1989) — `sarsa.py` was not covered by any of those PRs.
It is also one of the modules the elizaOS robot track imports in-process
(`core/{actor_critic,continual_backprop,initializers,normalizers,optimizers,
sarsa}`, per this repo's `CLAUDE.md`), so this is real external-facing
surface, not just an internal helper.

Both functions are exported public API (`alberta_framework.__init__`).

## Fix

Adds `_require_sarsa_sequence_length` (exact-type JAX-array + shape gate,
rejects length outside `[1, 10_000]` before any scan) and
`_require_sarsa_matching_length` (checks `rewards`/`terminated`/
`next_observations` share `observations`' leading length). The `10_000`
ceiling matches the documented learning-loop step ceiling already
established elsewhere in `core` (`learners._LEARNING_LOOP_MAX_STEPS`).

## Scope

- `alberta_framework/core/sarsa.py`
- `tests/test_sarsa.py`

## Verification

Commit under test: `1c564d2ba912a12dfab41add00ab4a8c8dbbe242`

```text
# On origin/main ec035f73 (pre-fix): no cap exists, jax.lax.scan is called
# unconditionally with the caller-supplied leading length -- confirmed by
# reading the pre-fix source (no _require_* call before jax.lax.scan in
# either function).

# At this head: oversized (10_001), origin-hang-class-sized (200_000), and
# length-mismatched inputs are rejected with a ValueError and a monkeypatch
# spy proves alberta_framework.core.sarsa.jax.lax.scan is never invoked.
# The documented last-fit length (10_000) still runs end-to-end.

.venv/bin/python -m pytest tests/test_sarsa.py -q -o addopts=""
# 149 tests collected (141 before this change, +8 new), 147 passed, 2 skipped

.venv/bin/python -m ruff check alberta_framework/core/sarsa.py tests/test_sarsa.py
# All checks passed!

.venv/bin/python -m mypy alberta_framework/core/sarsa.py
# Success: no issues found in 1 source file
.venv/bin/python -m mypy tests/test_sarsa.py
# 192 errors, all pre-existing on origin/main before this change (verified by
# diffing mypy output before/after with line numbers normalized); this PR
# introduces zero new mypy errors.

python -c "import alberta_framework"  # import surface stays intact
```

- [x] logs: local pytest/ruff/mypy output above; 149 collected / 147 passed / 2 skipped in tests/test_sarsa.py, zero new mypy errors vs. origin/main baseline

- [x] domain-artifact: N/A - scan sequence-length hang guard, no IPMNIST/benchmark shard produced

- AI assistance: yes

- Model(s) used: anthropic/claude-sonnet-5

- Agent tooling: Claude Code

- Skill path: contribute-to-asi (local skill copy)

- Provenance status: self-reported

AI provider/model: Anthropic / claude-sonnet-5
Client / agent tooling: Claude Code
Attribution status: self-reported

---

### CODE REVIEW

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.
